Make autopostgresqlbackup a bit more portable

dennisse opened the pull request:
GNU `stat` is not very portable. The less obvious, more portable and posix-friendly way to get the bytecount of a file is to use `wc(1)` - `wc -c < file`. This will work across all the Linuxes/unixes/BSDs.

`hostname` is also not very portable. Especially not with the `--fqdn`-flag. In fact, the man-page of [`hostname(1)`](https://manpages.debian.org/bookworm/hostname/hostname.1.en.html) kinda warns against using it.

The POSIX-friendly way of getting the hostname is `uname -n`, which will return the "node name". I believe `uname -n` uses the same function (`gethostname(2)`) `hostname(1)` uses. FQDN can be set either according to `hostname(1)`, or by `hostnamectl(1)` on  systemd systems.

dennisse committed:
Make autopostgresqlbackup a bit more portable

GNU `stat` is not very portable. The less obvious, more portable and
posix-friendly way to get the bytecount of a file is to use `wc(1)` -
`wc -c <
file`. This will work across all the Linuxes/unixes/BSDs.

`hostname` is also not very portable. Especially not with the
`--fqdn`-flag. Nn fact, the man-page of
[hostname(1)](https://manpages.debian.org/bookworm/hostname/hostname.1.en.html)
kinda warns against using it.

The POSIX-friendly way of getting the hostname is `uname -n`, which will
return the "node name". I believe `uname -n` uses the same
`gethostname(2)` which `hostname` uses. FQDN can be set either according
to hostname(1), or by hostnamectl(1) on systemd systems.

dennisse committed:
Just use printf to left-pad with zeroes

It's more portable and readable than sed

k0lter commented on autopostgresqlbackup:
```diff
@@ -161,7 +161,7 @@ DNOW="$(date '+%u')"                # Day number of the week 1 to 7 where 1 repr
 LOG_DIR="${BACKUPDIR}"              # Directory where the main log is saved
 # Fix day of month (left padding with 0)
-DOMONTHLY="$(echo "${DOMONTHLY}" | sed -r 's/^[0-9]$/0\0/')"
+DOMONTHLY="$(printf '%.2i' "${DOMONTHLY}")"
```
Fine!

k0lter commented on autopostgresqlbackup:
```diff
@@ -211,7 +211,7 @@ fi
 # pg_dump{,all} connection settings
 if [ "${DBHOST}" = "localhost" ]; then
-    HOST="$(hostname --fqdn)"
+    HOST="$(uname -n)"
```
I'd prefer to keep the FQDN hostname (instead of only the node name). We should find a portable way to get it. The use of `domainname` does not seems to be an option.

k0lter replied:
@dennisse I propose you to remove the hostname related changes and move them to another PR so that we could discuss the best to handle this. Once done I will merge your portability changes if you agree?

k0lter commented on autopostgresqlbackup:
```diff
@@ -315,7 +315,7 @@ dblist () {
     read -r -a dblist <<< "$(
         printf "%s" "${raw_dblist}" | \
-            sed -r -n 's/^([^:]+):.+$/\1/p' | \
+            sed -E -n 's/^([^:]+):.+$/\1/p' | \
```
Fine!

k0lter commented on autopostgresqlbackup:
```diff
@@ -444,7 +444,7 @@ dump() {
         log_debug "Fixing permissions (${PERM}) on '${dump_file}'"
         chmod "${PERM}" "${dump_file}"
-        fsize=$(stat -c '%s' "${dump_file}")
+        fsize=$(wc -c < "${dump_file}")
```
Fine!

k0lter commented on autopostgresqlbackup:
```diff
@@ -465,16 +465,10 @@ cleanup() {
     log_info "Rotating ${count} ${when} backups..."
     log_debug "Looking for '${db}_*' in '${dumpdir}/${when}/${db}'"
-    find "${dumpdir}/${when}/${db}/" -name "${db}_*" | \
-        sed -r 's/^.+([0-9]{4}-[0-9]{2}-[0-9]{2}_[0-9]{2}h[0-9]{2}m).*$/\1 \0/' | \
```
I failed to understand why you removed this part?

k0lter commented on autopostgresqlbackup:
```diff
@@ -566,8 +560,8 @@ if [ "${ENCRYPTION}" = "yes" ]; then
-            keyfp="$(echo "${keyinfo}" | sed -r -n 's/^\s*([a-z0-9]+)\s*$/\1/pi')"
-            keyuid="$(echo "${keyinfo}" | sed -r -n 's/^\s*uid\s+(\S.*)$/\1/pi' | head -n1)"
+            keyfp="$(echo "${keyinfo}" | sed -E -n 's/^\s*([a-z0-9]+)\s*$/\1/pi')"
+            keyuid="$(echo "${keyinfo}" | sed -E -n 's/^\s*uid\s+(\S.*)$/\1/pi' | head -n1)"
```
Fine!

k0lter reviewed (commented):
I added some comments on your changes.

dennisse committed:
`sed -r` is just GNU for POSIX `sed -E`

Many versions of `sed` allow for `-r` just to provide compatibility, but
`-E` is POSIX, and will be more portable.

dennisse committed:
`\0` is sugar added by GNU, and not very portable

However, `\1`, `\2`, etc, is POSIX, so just throwing an extra group
around the whole string should solve this.

k0lter commented:
@dennisse I've applied your portability improvements directly (see: https://github.com/k0lter/autopostgresqlbackup/commit/f341e5afce747bc97143641fb5a32631a6cb3c85). The only one I've discarded is the dump size calculation using `wc` which is in my opinion counter performant (with huge databases, It could take time to read the whole file). About `stat` portability, maybe we could replace it with find : `find FILE -printf '%s'`.

k0lter commented:
Thanks for you suggestion about `ls`. I tend to avoid spawning unneeded process so the first choice has my preference. 

About the hostname, I've already commit some kind of workaround, see: https://github.com/k0lter/autopostgresqlbackup/blob/master/autopostgresqlbackup#L202

nstanger commented:
BSD’s `stat` uses `-f` instead of `-c`. The following are equivalent (I have tested):

* Linux: `stat -c '%s'`
* BSD: `stat -f '%z'` (includes macOS/Darwin)

Of course then the problem becomes how to determine whether you are running on Linux or BSD (or something else). There are lots of ugly ways to do this, but Bash at least provides the `$OSTYPE` variable (e.g.,<https://stackoverflow.com/a/18434831>).

Even that is messy though, given that Darwin and BSD (others?) all effectively count as “BSD” for this purpose. (Although the script breaks on macOS anyway if you don’t install a modern Bash. Out of the box Apple **still** only provides Bash 3, which doesn’t have the `mapfile` builtin 🙄.)

JeroenKeizerNL commented:
kinda related, i tried running the script inside a alpine docker, the busybox version used in that also doesn't have the "hostname -fqdn" arguments, and also "xargs -L" seems to be missing

k0lter commented:
@JeroenKeizerNL I guess we could easily found a workaround for `hostname` and `--fqdn` unsupported option in busybox.
About `xargs`, there is the [findutils](https://pkgs.alpinelinux.org/package/edge/main/x86_64/findutils) package with `xargs` included. What do you think?

JeroenKeizerNL commented:
> @JeroenKeizerNL I guess we could easily found a workaround for `hostname` and `--fqdn` unsupported option in busybox. About `xargs`, there is the [findutils](https://pkgs.alpinelinux.org/package/edge/main/x86_64/findutils) package with `xargs` included. What do you think?

I saw the findutils version of xargs as well, that would probably work (didn't really test that one). Finding a workaround for the hostname --fqdn would be nice.  For my specific use case, i already build a container based on Debian, but being able to base it on Alpine should same a few MB on the image
